internal/multistorage/cache/status_cache.go

```diff
@@ -0,0 +1,160 @@
+package cache
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/wal-g/wal-g/pkg/storages/storage"
+)
+
+type StatusCache struct {
+	storages []NamedFolder
+	ttl      time.Duration
+}
+
+func NewStatusCache(primary storage.Folder, failover map[string]storage.Folder, ttl time.Duration) *StatusCache {
+	return &StatusCache{
+		storages: nameFolders(primary, failover),
+		ttl:      ttl,
+	}
+}
+
+func (c *StatusCache) AllAliveStorages() ([]NamedFolder, error) {
+	memCacheMu.Lock()
+	defer memCacheMu.Unlock()
+
+	if memCache.isRelevant(c.ttl, c.storages...) {
+		return memCache.getAllAlive(c.storages), nil
+	}
+
+	oldFile, err := readFile()
+	if err != nil {
+		return nil, fmt.Errorf("read cache file: %w", err)
+	}
+	_, outdated := oldFile.splitByRelevance(c.ttl, c.storages)
+	if len(outdated) == 0 {
+		memCache = oldFile
+		return memCache.getAllAlive(c.storages), nil
+	}
+
+	checkResult, err := checkForAlive(outdated...)
+	if err != nil {
+		return nil, fmt.Errorf("find alive storages: %w", err)
+	}
+
+	newFile := updateFileContent(oldFile, checkResult)
+	err = writeFile(newFile)
+	if err != nil {
+		return nil, fmt.Errorf("write cache file: %w", err)
+	}
+
+	memCache = newFile
+	return memCache.getAllAlive(c.storages), nil
+}
+
+func (c *StatusCache) FirstAliveStorage() (*NamedFolder, error) {
```

Comment:
We still need to retain the logic where "default" storage is chosen if it is available. See the [current implementation](https://github.com/wal-g/wal-g/blob/master/internal/multistorage/alive_checker.go#L16) for details.

Reply:
It already works so. For the same reason as in [the comment](https://github.com/wal-g/wal-g/pull/1514#discussion_r1273831782)

I'll rename something in my code and/or add some comments to make it clearer.